### ISSUE TITLE
Support IAM roles for service accounts (new CredentialsProvider)

### ISSUE THREAD

**Issue by 0xpablo** [swift-aws/aws-sdk-swift-core#300]
**Describe the solution you'd like**
It would be great to support IAM roles for service accounts for authentication.
Here's the [AWS Documentation](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html).
This is very useful for customizing credentials down to the pod level in EKS clusters (so that the credentials aren't shared between Pods and so that they can have the least privilege possible)

**Additional context**
This is AWS' Java implementation:
[WebIdentityTokenFileCredentialsProvider.java](https://github.com/aws/aws-sdk-java-v2/blob/master/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/WebIdentityTokenFileCredentialsProvider.java)
[StsWebIdentityCredentialsProviderFactory.java](https://github.com/aws/aws-sdk-java-v2/blob/master/services/sts/src/main/java/software/amazon/awssdk/services/sts/internal/StsWebIdentityCredentialsProviderFactory.java)

**Comment by adam-fowler** [comment]
Don't know a great deal about EKS, but you can use STS direct to provide web identity credentials for you. I believe the following would work.
```
import AWSSTS

// create STS client with empty credentials
let sts = STS(accessKeyId: "", secretAccessKey: "")

let request = STS.AssumeRoleWithWebIdentityRequest(roleArn: "arn:aws:iam::000000000000:role/myRole", webIdentityToken: "myWebToken")
let response = try sts.assumeRoleWithWebIdentity(request).wait()
let credentials = response.credentials
```
If you are using the 5.0.0 alpha we are almost at the point where you can supply your own CredentialProvider to `AWSClient`. Once that is in it wouldn't be too hard to add an extension to STS that is a `WebIdentityCredentialProvider` that wraps the above code. One thing that would need to be resolved is what happens when the provided credentials expire, do we ask for new credentials and when we do we have to be sure the `webIdentityToken` hasn't also expired.

**Comment by adam-fowler** [comment]
@0xpablo if we had a AssumeRoleWithWebIdentity credential provider would you want it to re-new when it is about to expire? In this situation I'm assuming you would need to provide me with a new web identity or should the existing one work ok.

**Comment by adam-fowler** [comment]
@0xpablo https://github.com/swift-aws/aws-sdk-swift/pull/333

**Comment by adam-fowler** [comment]
@0xpablo are you in a position to test https://github.com/swift-aws/aws-sdk-swift#333? Would be interested to hear your feedback

**Comment by adam-fowler** [comment]
PR is now merged and code is available in master. See docs https://github.com/swift-aws/aws-sdk-swift/blob/5.x.x-docs/documentation/credentials.md#sts-and-cognito-identity

**Comment by 0xpablo** [comment]
Thank you for this @adam-fowler! I'm sorry that I haven't been able to test this earlier. I have updated the library to the latest alpha and removed the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` env vars I was using as a workaround and it seems to be working well 👍 

```

2020-09-17T09:06:54+0000 info com.cor.AWS : aws-credential-provider=RotatingCredentialProvider(InstanceMetaDataClient) Refeshing AWS credentials
2020-09-17T09:06:54+0000 debug com.cor.AWS : ahc-channel-el=SelectableEventLoop { selector = Selector { descriptor = 3 }, thread = NIOThread(name = NIO-ELT-0-#0) } ahc-request-id=0 ahc-connection=SocketChannel { BaseSocket { fd=12 }, active = true, localAddress = Optional([IPv4]192.168.120.61/192.168.120.61:45470), remoteAddress = Optional([IPv4]169.254.169.254/169.254.169.254:80) } ahc-request=PUT http://169.254.169.254/latest/api/token ahc-task-el=SelectableEventLoop { selector = Selector { descriptor = 3 }, thread = NIOThread(name = NIO-ELT-0-#0) } got connection for request
2020-09-17T09:06:56+0000 debug com.cor.AWS : ahc-request-id=0 ahc-closing=true releasing connection, request complete
2020-09-17T09:06:56+0000 debug com.cor.AWS : ahc-request-id=0 ahc-provider=AsyncHTTPClient.HTTP1ConnectionProvider closing provider
2020-09-17T09:06:56+0000 info com.cor.AWS : Did not find IMDSv2 token, use IMDSv1
2020-09-17T09:06:56+0000 debug com.cor.AWS : ahc-channel-el=SelectableEventLoop { selector = Selector { descriptor = 3 }, thread = NIOThread(name = NIO-ELT-0-#0) } ahc-request-id=1 ahc-request=GET http://169.254.169.254/latest/meta-data/iam/security-credentials/ ahc-connection=SocketChannel { BaseSocket { fd=12 }, active = true, localAddress = Optional([IPv4]192.168.120.61/192.168.120.61:45500), remoteAddress = Optional([IPv4]169.254.169.254/169.254.169.254:80) } ahc-task-el=SelectableEventLoop { selector = Selector { descriptor = 3 }, thread = NIOThread(name = NIO-ELT-0-#0) } got connection for request
2020-09-17T09:06:56+0000 debug com.cor.AWS : ahc-request-id=1 ahc-closing=true releasing connection, request complete
2020-09-17T09:06:56+0000 debug com.cor.AWS : ahc-request-id=1 ahc-provider=AsyncHTTPClient.HTTP1ConnectionProvider closing provider
2020-09-17T09:06:56+0000 debug com.cor.AWS : ahc-connection=SocketChannel { BaseSocket { fd=13 }, active = true, localAddress = Optional([IPv4]192.168.120.61/192.168.120.61:45502), remoteAddress = Optional([IPv4]169.254.169.254/169.254.169.254:80) } ahc-request-id=2 ahc-request=GET http://169.254.169.254/latest/meta-data/iam/security-credentials/*** ahc-channel-el=SelectableEventLoop { selector = Selector { descriptor = 3 }, thread = NIOThread(name = NIO-ELT-0-#0) } ahc-task-el=SelectableEventLoop { selector = Selector { descriptor = 3 }, thread = NIOThread(name = NIO-ELT-0-#0) } got connection for request
2020-09-17T09:06:56+0000 debug com.cor.AWS : ahc-request-id=2 ahc-closing=true releasing connection, request complete
2020-09-17T09:06:56+0000 debug com.cor.AWS : ahc-request-id=2 ahc-provider=AsyncHTTPClient.HTTP1ConnectionProvider closing provider
2020-09-17T09:06:56+0000 info com.cor.AWS : aws-credential-provider=RotatingCredentialProvider(InstanceMetaDataClient) AWS credentials ready
2020-09-17T09:06:56+0000 info com.cor.AWS : aws-credential-provider=RotatingCredentialProvider(InstanceMetaDataClient) Select credential provider
```

**Comment by adam-fowler** [comment]
@0xpablo little confused here. The log you posted indicates you are acquiring credentials via ec2 instance metadata which has always been available. You didnt need the STSAssumeRoleWithWebIdentity credential provider? No worries your issue pushed me to ensure the CredentialProvider system was flexible enough to support CredentialProviders outside of the core and then implement the STS and Cognito providers.

**Comment by 0xpablo** [comment]
Oh, I've been reviewing the history of the repo and the available credential providers when I opened the ticket and I think I overlooked the ECS one... I thought that wasn't available so I injected the variables that I mentioned before, but I think it would have worked. I didn't know AWS exposed that API (the one with the hardcoded IP) to get credentials on ECS containers.

I think the confusion came because initially I was looking for the `AWS_ROLE_ARN` and `AWS_WEB_IDENTITY_TOKEN_FILE` strings on the repo. As those are the environment variables that AWS populates for ECS containers, and that's what our Java services are using.
I was assuming that that's what aws-sdk-swift#333 was doing, but I was surprised that the PR wasn't on the core repo. I understand the `CognitoIdentity.CredentialsProvider` and I think it will be useful for some people, but that's not exactly what I was after.
I was after a core credentials provider (that was actually part of the default chain of providers) that looked into those env vars and grabbed the token from there.
Anyway I overlooked the ecs provider that uses the instance metadata API which also works. I'm not an expert on AWS so I don't know if supporting the `AWS_WEB_IDENTITY_TOKEN_FILE` makes sense or not and if it should take precedence over the ecs provider.

I don't want to take more of your time, I hope I have some extra time in a few weeks and if no one has implemented that provider I might try to do it and submit a PR.

Thanks for your time @adam-fowler
